### PR TITLE
Show a list of vacancy applicants

### LINKED ISSUE #9: Vacancies - Applications
- [ ] show applications
- [ ] accept or deny an application

### PR DESCRIPTION
**Pull request by DeKleineKobini**
Closes #9

### CODE REVIEW

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            @PathVariable("id") long id,
+            @Valid @RequestBody ApplicationUpdateStatusDto dto
+    ) {
+        if (dto.getId() != id) {
+            throw new EndpointException(GlobalExceptionMessage.ID_MISMATCH);
+        }
+
+        Application application = applicationService.findById(id)
+                .orElseThrow(() -> new EndpointException(HttpStatus.NOT_FOUND, GlobalExceptionMessage.NOT_FOUND));
+
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java]: Shoudn't this be done in the service?

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/BaseController.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+public abstract class BaseController {
+
+    protected String getCurrentEmail() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        return authentication == null ? null : authentication.getName();
+    }
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/BaseController.java]: Maybe just do this in the securityservice

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        this.vacancyService = vacancyService;
+        this.applicationService = applicationService;
+        this.applicationMapper = applicationMapper;
+    }
+
+    @PutMapping("{id}/status")
+    @RolesAllowed("RECRUITER")
+    public ApplicationReadDto updateApplicationStatus(
+            @PathVariable("id") long id,
+            @Valid @RequestBody ApplicationUpdateStatusDto dto
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java]: Why provide id twice?

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/services/VacancyService.java [hunk 4]
```diff
@@ -42,4 +46,23 @@ public Vacancy addVacancy(Vacancy vacancy) {
...
+            hasAccess = vacancy.getRecruiter().getEmail().equals(email);
+        } else if (vacancy.getCompany() != null) {
+            Optional<Account> accountOptional = accountService.findByEmail(email);
+
+            if (accountOptional.isPresent()
+                    && accountOptional.get() instanceof Recruiter recruiter
+                    && recruiter.getCompany() != null) {
+                hasAccess = vacancy.getCompany().getId().equals(recruiter.getCompany().getId());
+            }
+        }
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/services/VacancyService.java:74]: So if the vacancy has a linked recruiter, the others in the company can't access the vacancy?

File: inetum-realjobs-backend/src/main/resources/data-postgres.sql [hunk 5]
```diff
@@ -1,512 +1,553 @@
...
+                            password, profile_picture, country_id)
+VALUES (2, 'initial', '2021-12-28 14:21:14.000000', 'initial', '2021-12-28 14:21:16.000000', 0, null, 'Brussels', '7',
+        '1000', 'Road', '2021-12-01', 'recruiter@freelance.com', 'User', 'MALE', 'User', null,
+        '$2a$10$wrvu9zLKCX7zFtaLBFNVKeBh8Y..TLZCuvnzkUcL7N7/OjTD.ximq', null, 1);
+ALTER SEQUENCE account_id_seq RESTART WITH 4;
+
+INSERT INTO public.job_seeker (id, resume_id)
+VALUES (1, null);
+INSERT INTO public.recruiter (id, company_id)
+VALUES (2, null);
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/main/resources/data-postgres.sql]: Resume will be made mandatory in #49, please add resumes here (they can be empty)

File: inetum-realjobs-frontend/src/app/services/application.service.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+@Injectable({
+  providedIn: "root",
+})
+export class ApplicationService {
+
+  constructor(private httpClient: HttpClient) {
+  }
+
+  updateApplication(id: number, status: ApplicationStatus): Observable<Application> {
+    return this.httpClient.put<Application>(`http://localhost:8080/api/applications/${id}/status`, {id, status});
```
Comment (TomD011099) [inline inetum-realjobs-frontend/src/app/services/application.service.ts]: Same as controller, why mention id twice?

File: inetum-realjobs-backend/src/main/resources/application.properties [hunk 7]
```diff
@@ -21,6 +21,8 @@ application.mail.from=noreply@realjobs.com
 spring.jpa.properties.javax.persistence.schema-generation.scripts.action=create
 spring.jpa.properties.javax.persistence.schema-generation.scripts.create-target=create-tables.sql
 spring.jpa.properties.javax.persistence.schema-generation.scripts.create-source=metadata
+spring.jpa.properties.hibernate.hbm2ddl.schema-generation.script.append=false
+spring.jackson.serialization.FAIL_ON_EMPTY_BEANS=false
```
Comment (TomD011099) [inline inetum-realjobs-backend/src/main/resources/application.properties]: What's this one for?

**Review by TomD011099 — CHANGES_REQUESTED** [review]

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java (hunk 1 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java]: I always check my security in the controller, and I need this application for the security check. Can be changed if you wish to do so.

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java (hunk 3 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/ApplicationController.java]: That's how REST specifies it and it became an habit, no other reason. Can be changed if you wish to do so.

File: inetum-realjobs-frontend/src/app/services/application.service.ts (hunk 6 above)
Comment (DeKleineKobini) [reply inetum-realjobs-frontend/src/app/services/application.service.ts]: That's how REST specifies it and it became an habit, no other reason. Can be changed if you wish to do so.

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/services/VacancyService.java (hunk 4 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/services/VacancyService.java:74]: How I understood how vacancies should work there are 2 possibilities:

- a vacancy is linked to a recruiter
- a vacancy is linked to a company

Either recruiter is null, or company is null.

File: inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/BaseController.java (hunk 2 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/main/java/world/inetum/realdolmen/realjobs/controllers/BaseController.java]: I always check my security in the controller, including information that is gotten from the context. Can be changed if you wish to do so.

File: inetum-realjobs-backend/src/main/resources/application.properties (hunk 7 above)
Comment (DeKleineKobini) [reply inetum-realjobs-backend/src/main/resources/application.properties]: I had something fail due to an empty bean error. Might not be needed anymore.

**Review by TomD011099 — APPROVED** [review]

**Review by redasaada — APPROVED** [review]

**Review by JohanSchoubben — APPROVED** [review]

**Pull request merged**